src/10_fixed_functions.rs

```diff
@@ -0,0 +1,438 @@
+#[macro_use]
+extern crate lazy_static;
+extern crate glsl_to_spirv;
+
+use hal::{Capability, Device, Instance, PhysicalDevice, QueueFamily, Surface, SwapchainConfig};
+use std::{fs, io::Read};
+
+static WINDOW_NAME: &str = "10_fixed_functions";
+
+fn main() {
+    log_output_setup();
+    let (window, events_loop) = init_window();
+    // we must return surface, since it must be destroyed before swapchain
+    let (
+        _instance,
+        device,
+        swapchain,
+        _surface,
+        frame_images,
+        descriptor_set_layouts,
+        pipeline_layout,
+    ) = init_hal(&window);
+    main_loop(events_loop);
+    // must destroy the swapchain
+    clean_up(
+        device,
+        frame_images,
+        swapchain,
+        descriptor_set_layouts,
+        pipeline_layout,
+    );
+}
+
+lazy_static! {
+    static ref VERT_SHADER_CODE: Vec<u8> = {
+        let glsl = fs::read_to_string("src/09_shader_base.vert")
+            .expect("Error loading vertex shader code.");
+        glsl_to_spirv::compile(&glsl, glsl_to_spirv::ShaderType::Vertex)
+            .expect("Error compiling vertex shader code.")
+            .bytes()
+            .map(|b| b.unwrap())
+            .collect::<Vec<u8>>()
+    };
+    static ref FRAG_SHADER_CODE: Vec<u8> = {
+        let glsl = fs::read_to_string("src/09_shader_base.frag")
+            .expect("Error loading fragment shader code.");
+        glsl_to_spirv::compile(&glsl, glsl_to_spirv::ShaderType::Fragment)
+            .expect("Error compiling fragment shader code.")
+            .bytes()
+            .map(|b| b.unwrap())
+            .collect::<Vec<u8>>()
+    };
+}
+
+fn create_graphics_pipeline(
+    device: &<back::Backend as hal::Backend>::Device,
+    extent: &hal::window::Extent2D,
+) -> (
+    Vec<<back::Backend as hal::Backend>::DescriptorSetLayout>,
+    <back::Backend as hal::Backend>::PipelineLayout,
+) {
+    let vert_shader_module = device
+        .create_shader_module(&VERT_SHADER_CODE)
+        .expect("Error creating shader module.");
+    let frag_shader_module = device
+        .create_shader_module(&FRAG_SHADER_CODE)
+        .expect("Error creating fragment module.");
+
+    let (ds_layouts, pipeline_layout) = {
+        let (vs_entry, fs_entry) = (
+            hal::pso::EntryPoint::<back::Backend> {
+                entry: "main",
+                module: &vert_shader_module,
+                specialization: &[],
+            },
+            hal::pso::EntryPoint::<back::Backend> {
+                entry: "main",
+                module: &frag_shader_module,
+                specialization: &[],
+            },
+        );
+
+        // shader stage set up
+        let shaders = hal::pso::GraphicsShaderSet {
+            vertex: vs_entry,
+            hull: None,
+            domain: None,
+            geometry: None,
+            fragment: Some(fs_entry),
+        };
+
+        // rasterizer: is there a field to set the rasterizer discard option?
```

Comment:
In the current HAL implementation, this is derived based on whether a fragment shader was provided. We can probably just leave this out for now.